packages/playwright/src/plugins/webServerPlugin.ts

```diff
@@ -102,14 +119,35 @@ export class WebServerPlugin implements TestRunnerPlugin {
       cwd: this._options.cwd,
       stdio: 'stdin',
       shell: true,
-      // Reject to indicate that we cannot close the web server gracefully
-      // and should fallback to non-graceful shutdown.
-      attemptToGracefullyClose: () => Promise.reject(),
+      attemptToGracefullyClose: async () => {
+        if (process.platform === 'win32')
+          throw new Error('Graceful shutdown is not supported on Windows');
+        if (!signal)
+          throw new Error('skip graceful shutdown');
+
+        const success = launchedProcess.kill(signal);
+        if (!success)
+          throw new Error(`SIGINT didn't succeed, fall back to non-graceful shutdown`);
+
+        const processExit = new Promise<void>(f => launchedProcess.once('exit', f));
+        if (timeout === 0) {
+          await processExit;
+        } else {
+          await Promise.race([
+            processExit,
```

Comment:
If process exits before the timer runs out, shouldn't we clear the timer? If we don't, it would probably block Node.js from exiting for the full duration.

Reply:
good catch, yeah! I've refactored this a tad bit to make it easier on the eye. Noticed that `node:timers` has some useful tools around this though, particular `ref` and `signal`: https://nodejs.org/api/timers.html#timerspromisessettimeoutdelay-value-options
Might be handy somewhere else.